niceman/interface/exec.py

```diff
@@ -32,6 +33,211 @@
 lgr = getLogger('niceman.api.exec')
 
 
+class CommandAdapter(object):
+    """Interface to kinds of commands `exec` can execute.
+
+    Each subclass should define a `execute` method and optionally defines
+    `pre_command` or `post_command` to perform pre- or post-command operations.
+    The caller should execute the command by calling the instance object
+    itself.
+
+    Parameters
+    ----------
+    resource : Resource object
+    command : str
+        The program to run.
+    cmd_args : list of str
+        Arguments to pass to `command`.
+    """
+
+    def __init__(self, resource, command, cmd_args):
+        self.resource = resource
+        self.session = resource.get_session()
+        self.command = command
+        self.cmd_args = cmd_args
+
+    def pre_command(self):
+        pass
+
+    def execute(self):
+        """Run the full command defined by `command` and `cmd_args`.
+
+        Returns
+        -------
+        A tuple, (stdout, stderr).
+        """
+        raise NotImplementedError
```

Comment:
we do use sparingly the `@abc.abstractmethod` etc sparingly through out the code, might be a good candidate to use them here as well

Reply:
Yes, I know, because it's on my to-do list to fix up how they are used in niceman :]  In Python 3, they don't raise the expected type error when a creating a child with unimplemented methods.  This is because the `__metaclass__ = abc.ABCMeta` isn't checked in Python 3 (which expects a `metaclass` keyword argument instead).  There's also the cosmetic issue that many of the methods marked with `abstractmethod` raise a `NotImplementedError`, which is pointless.

But for this specific case, I don't think the complication of using an ABC is worth it (especially given the py2/py3 compatibility kludge) because these classes are really just for internal organization of `exec.py`.